msgpack/src/main/scala/fs2/data/msgpack/low/internal/ItemSerializer.scala

```diff
@@ -0,0 +1,247 @@
+      *
+      * Use this instead of [[Out.push]] when `bv` may significantly exceed 4KiB.
+      */
+    def pushBuffered(bv: ByteVector): Pull[F, Byte, Out[F]] = {
+      @inline
+      def go(chunk: Chunk[Byte], rest: ByteVector): Pull[F, Byte, Out[F]] =
+        if (rest.isEmpty)
+          Pull.done.as(new Out(chunk))
+        else
+          Pull.output(chunk) >> go(Chunk.byteVector(rest.take(limit.toLong)), rest.drop(limit.toLong))
+
+      if (bv.isEmpty)
+        this.push(bv)
+      else if (contents.size >= limit)
+        Pull.output(contents) >> go(Chunk.byteVector(bv.take(limit.toLong)), bv.drop(limit.toLong))
+      else
+        go(contents ++ Chunk.byteVector(bv.take(limit.toLong - contents.size)), bv.drop(limit.toLong - contents.size))
+    }
+
+    /** Outputs the whole buffer. */
+    @inline
+    def flush = Pull.output(contents)
+  }
+
+  @inline
+  private def step[F[_]: RaiseThrowable](o: Out[F], item: MsgpackItem): Pull[F, Byte, Out[F]] = item match {
+    case MsgpackItem.UnsignedInt(bytes) =>
+      val bs = bytes.dropWhile(_ == 0)
+      if (bs.size <= 1)
+        o.push(ByteVector(Headers.Uint8) ++ bs.padLeft(1))
+      else if (bs.size <= 2)
+        o.push(ByteVector(Headers.Uint16) ++ bs.padLeft(2))
+      else if (bs.size <= 4)
+        o.push(ByteVector(Headers.Uint32) ++ bs.padLeft(4))
+      else if (bs.size <= 8)
+        o.push(ByteVector(Headers.Uint64) ++ bs.padLeft(8))
+      else
+        Pull.raiseError(new MalformedUintError)
+
+    case MsgpackItem.SignedInt(bytes) =>
+      val bs = bytes.dropWhile(_ == 0)
+      if (bs.size <= 1)
+        // positive fixint or negative fixint
+        if ((bs & hex"7f") == bs || (bs & hex"c0") == hex"c0")
+          o.push(bs.padLeft(1))
+        else
+          o.push(ByteVector(Headers.Int8) ++ bs.padLeft(1))
+      else if (bs.size <= 2)
+        o.push(ByteVector(Headers.Int16) ++ bs.padLeft(2))
+      else if (bs.size <= 4)
+        o.push(ByteVector(Headers.Int32) ++ bs.padLeft(4))
+      else if (bs.size <= 8)
+        o.push(ByteVector(Headers.Int64) ++ bs.padLeft(8))
+      else
+        Pull.raiseError(new MalformedIntError)
+
+    case MsgpackItem.Float32(float) =>
+      o.push(ByteVector(Headers.Float32) ++ ByteVector.fromInt(java.lang.Float.floatToIntBits(float)))
+
+    case MsgpackItem.Float64(double) =>
+      o.push(ByteVector(Headers.Float64) ++ ByteVector.fromLong(java.lang.Double.doubleToLongBits(double)))
+
+    case MsgpackItem.Str(bytes) =>
+      if (bytes.size <= 31) {
+        o.push(ByteVector.fromByte((0xa0 | bytes.size).toByte) ++ bytes)
+      } else if (bytes.size <= Math.pow(2, 8) - 1) {
+        val size = ByteVector.fromByte(bytes.size.toByte)
+        o.push(ByteVector(Headers.Str8) ++ size ++ bytes)
+      } else if (bytes.size <= Math.pow(2, 16) - 1) {
+        val size = ByteVector.fromShort(bytes.size.toShort)
+        o.push(ByteVector(Headers.Str16) ++ size ++ bytes)
+      } else if (fitsIn(bytes.size.toInt, 32)) {
```

Comment:
Isn't this guaranteed to be `true`? `.toInt` will truncate the size to `2 ^31 - 1` which always fits in 32 bits, doesn't it? 🤔 

Also, while it's cool to have a generic function here, just comparing the size to `Int.MaxValue` might be much more efficient.